banyan/src/forest/read.rs

```diff
@@ -19,7 +13,288 @@ use anyhow::{anyhow, Result};
 use core::fmt::Debug;
 use futures::{prelude::*, stream::BoxStream};
 use libipld::cbor::DagCbor;
+use smallvec::{smallvec, SmallVec};
+
 use std::{iter, sync::Arc, time::Instant};
+enum Mode {
+    Forward,
+    Backward,
+}
+pub(crate) struct ForestIter<T: TreeTypes, V, R, Q: Query<T>, F> {
+    forest: Forest<T, V, R>,
+    offset: u64,
+    query: Q,
+    mk_extra: F,
+    index_stack: SmallVec<[Arc<Index<T>>; 64]>,
```

Comment:
64 is probably a bit on the high side for this. For a packed tree, you don't need more than 5. And as soon as you are traversing a really deep chain, allocating here is the least of your problems.